MAJI NDOGO PROJECT 

Integrated project: 

Maji Ndogo is a country facing water related problem. The president of the country; Aziza Naledi has contracted some field workers and a Data Analytics Company to collect data from the people using the various types of water and to help her come up with data-based solutions.

The survey has come to a successful end.

Beginning our data-driven journey in Maji Ndogo (Part 1)
In this first part of the integrated project, I will dive into Maji Ndogo's expansive database containing 60,000 records spread across various tables. As I navigate this trove of data, I will use basic queries to familiarise myself with the content of each table. Along the way, I will also refine some data using DML(Data Manupulation Language).

Connecting to MySQL database 

In [1]:
# Loading and activating the SQL extention to allow the execution of SQL in a Jupyter notebook

%load_ext sql

In [ ]:
# Establishing a connection to the local database

%sql mysql+pymysql://root:password@localhost:3306/md_water_services


'Connected: root@md_water_services'

Getting to Know my datasets 

In [3]:
%%sql

SHOW TABLES;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
9 rows affected.


[('data_dictionary',),
 ('employee',),
 ('global_water_access',),
 ('location',),
 ('visits',),
 ('water_quality',),
 ('water_source',),
 ('well_pollution',),
 ('well_pollution_copy',)]

In [4]:
%%sql

# Fetching records from the location table

SELECT
    *
FROM 
    location
LIMIT 10;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
10 rows affected.


[('AkHa00000', '2 Addis Ababa Road', 'Akatsi', 'Harare', 'Urban'),
 ('AkHa00001', '10 Addis Ababa Road', 'Akatsi', 'Harare', 'Urban'),
 ('AkHa00002', '9 Addis Ababa Road', 'Akatsi', 'Harare', 'Urban'),
 ('AkHa00003', '139 Addis Ababa Road', 'Akatsi', 'Harare', 'Urban'),
 ('AkHa00004', '17 Addis Ababa Road', 'Akatsi', 'Harare', 'Urban'),
 ('AkHa00005', '125 Addis Ababa Road', 'Akatsi', 'Harare', 'Urban'),
 ('AkHa00006', '98 Addis Ababa Road', 'Akatsi', 'Harare', 'Urban'),
 ('AkHa00007', '21 Addis Ababa Road', 'Akatsi', 'Harare', 'Urban'),
 ('AkHa00008', '11 Addis Ababa Road', 'Akatsi', 'Harare', 'Urban'),
 ('AkHa00009', '6 Addis Ababa Road', 'Akatsi', 'Harare', 'Urban')]

Task 2: Dive into the water sources

In [5]:
%%sql 

SELECT
    *
FROM 
    water_source
LIMIT 5;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[('AkHa00000224', 'tap_in_home', 956),
 ('AkHa00001224', 'tap_in_home_broken', 930),
 ('AkHa00002224', 'tap_in_home_broken', 486),
 ('AkHa00003224', 'well', 364),
 ('AkHa00004224', 'tap_in_home_broken', 942)]

In [6]:
%%sql 

# Taking a look at the Data dictionary table to get more info about the tables and the various columns 

SELECT 
    *
FROM 
    data_dictionary

 * mysql+pymysql://root:***@localhost:3306/md_water_services
49 rows affected.


[('employee', 'assigned_employee_id', 'Unique ID assigned to each employee', 'INT', 'visits'),
 ('employee', 'employee_name', 'Name of the employee', 'VARCHAR(255)', ''),
 ('employee', 'phone_number', 'Contact number of the employee', 'VARCHAR(15)', ''),
 ('employee', 'email', 'Email address of the employee', 'VARCHAR(255)', ''),
 ('employee', 'address', 'Residential address of the employee', 'VARCHAR(255)', ''),
 ('employee', 'town_name', 'Name of the town where the employee resides', 'VARCHAR(255)', ''),
 ('employee', 'province_name', 'Name of the province where the employee resides', 'VARCHAR(255)', ''),
 ('employee', 'position', 'Position or job title of the employee', 'VARCHAR(255)', ''),
 ('visits', 'record_id', 'Unique ID assigned to each visit', 'int', 'water_quality, water_source'),
 ('visits', 'location_id', 'ID of the location visited', 'varchar(255)', 'location'),
 ('visits', 'source_id', 'ID of the water source visited', 'varchar(510)', 'well_pollution'),
 ('visits', 'time_of_record', 'Date and time of the visit', 'datetime', ''),
 ('visits', 'visit_count', 'Number of visits made to this location', 'int', ''),
 ('visits', 'time_in_queue', 'Time spent by people waiting for water in a queue at the location', 'int', ''),
 ('visits', 'assigned_employee_id', 'ID of the employee who visited the location', 'int', 'employee'),
 ('water_quality', 'record_id', 'Unique ID assigned to each record', 'int', ''),
 ('water_quality', 'subjective_quality_score', 'Score representing the subjective quality of the water source', 'int', ''),
 ('water_quality', 'visit_count', 'Number of visits made for data collection', 'int', ''),
 ('water_source', 'source_id', 'Unique ID assigned to each water source', 'VARCHAR(510)', 'visits'),
 ('water_source', 'type_of_water_source', 'Type or category of the water source. Can be : tap_in_home, tap_in_home_broken, well, shared_tap, river', 'VARCHAR(255)', ''),
 ('water_source', 'number_of_people_served', 'Number of people served by this water source', 'INT', ''),
 ('well_pollution', 'source_id', 'ID of the water source being tested for pollution', 'varchar(258)', 'visits'),
 ('well_pollution', 'date', 'Date of the pollution test', 'datetime', ''),
 ('well_pollution', 'description', 'Description of the pollution test', 'varchar(255)', ''),
 ('well_pollution', 'pollutant_ppm', 'Result of the pollution test in parts per million', 'float', ''),
 ('well_pollution', 'biological', 'Biological contamination level', 'float', ''),
 ('well_pollution', 'results', "Result of the pollution test. Can be 'dirty', 'biologically contaminated', or 'clean'", 'varchar(255)', ''),
 ('location', 'location_id', 'Unique ID assigned to each location', 'varchar(255)', 'visits'),
 ('location', 'address', 'Address of the location', 'varchar(255)', ''),
 ('location', 'province_name', 'Name of the province where the location is situated', 'varchar(255)', ''),
 ('location', 'town_name', 'Name of the town where the location is situated', 'varchar(255)', ''),
 ('location', 'location_type', 'Type or category of the location', 'varchar(255)', ''),
 ('global_water_access', 'name', 'The country or area name', 'varchar(255)', ''),
 ('global_water_access', 'region', 'Geographical region', 'varchar(255)', ''),
 ('global_water_access', 'year', 'Year of the data record', 'int', ''),
 ('global_water_access', 'pop_n', 'The national population size estimate in thousands', 'float', ''),
 ('global_water_access', 'pop_u', 'The urban population share estimate in percentage points (%)', 'float', ''),
 ('global_water_access', 'wat_bas_n', 'The estimated national share of people with at least basic service (%)', 'float', ''),
 ('global_water_access', 'wat_lim_n', 'The estimated national share of people with limited service (%)', 'float', ''),
 ('global_water_access', 'wat_unimp_n', 'The estimated national share of people with unimproved service (%)', 'float', ''),
 ('global_water_access', 'wat_sur_n', 'The estimated national share of people with surfa

Diving into the water sources
We need to understand the types of water sources we're dealing with.

In [7]:
%%sql

#   querying the unique types of water resources
     
SELECT 
    DISTINCT(type_of_water_source)
FROM 
    water_source;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[('tap_in_home',),
 ('tap_in_home_broken',),
 ('well',),
 ('shared_tap',),
 ('river',)]

The various source types and a little info on them:
1. *River* - People collect drinking water along a river. This is an open water source that millions of people use in Maji Ndogo. Water from a river has a high risk of being contaminated with biological and other pollutants, so it is the worst source of water possible.

2. *Well* - These sources draw water from underground sources, and are commonly shared by communities. Since these are closed water sources, contamination is much less likely compared to a river. Unfortunately, due to the aging infrastructure and the corruption of officials in the past, many of our wells are not clean.

3. *Shared tap* - This is a tap in a public area shared by communities.

4. *Tap in home* - These are taps that are inside the homes of our citizens. On average about 6 people live together in Maji Ndogo, so
each of these taps serves about 6 people.

5. *Broken tap in home* - These are taps that have been installed in a citizen’s home, but the infrastructure connected to that tap is not functional. This can be due to burst pipes, broken pumps or water treatment plants that are not working.

*An important note on the home taps*: About 6-10 million people have running water installed in their homes in Maji Ndogo, including
broken taps. If we were to document this, we would have a row of data for each home, so that one record is one tap. That means the
database would contain about 1 million rows of data, which may slow the systems down. For now, the surveyors combined the data of
many households together into a single record.

Task 3: Unpack visits to water sources

In [8]:
%%sql 

SELECT 
    *
FROM 
    visits
LIMIT 5;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[(0, 'SoIl32582', 'SoIl32582224', datetime.datetime(2021, 1, 1, 9, 10), 1, 15, 12),
 (1, 'KiRu28935', 'KiRu28935224', datetime.datetime(2021, 1, 1, 9, 17), 1, 0, 46),
 (2, 'HaRu19752', 'HaRu19752224', datetime.datetime(2021, 1, 1, 9, 36), 1, 62, 40),
 (3, 'AkLu01628', 'AkLu01628224', datetime.datetime(2021, 1, 1, 9, 53), 1, 0, 1),
 (4, 'AkRu03357', 'AkRu03357224', datetime.datetime(2021, 1, 1, 10, 11), 1, 28, 14)]

Task 4: Assess the quality of water sources
The quality of our water sources is the whole point of this survey.

In [9]:
%%sql 

SELECT 
    *
FROM 
    water_quality
LIMIT 5;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[(0, 0, 1), (1, 1, 1), (2, 5, 1), (3, 10, 1), (4, 4, 1)]

The surveyors only made multiple visits to shared taps and did not revisit other types of water sources. So
there should be no records of second visits to locations where there are good water sources, like taps in homes. 
Lets see if this is true. So I am going to query out tap_in_home sources where visits = 2 to see if i will indeed not find a data.

In [10]:
%%sql

SELECT 
    ws.type_of_water_source,
    wq.subjective_quality_score,
    v.visit_count
FROM 
    water_source ws
JOIN 
    visits v ON ws.source_id = v.source_id
JOIN 
    water_quality wq ON v.record_id = wq.record_id
WHERE 
    wq.subjective_quality_score = 10
    AND v.visit_count = 2; 

 * mysql+pymysql://root:***@localhost:3306/md_water_services
218 rows affected.


[('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 10, 2),
 ('shared_tap', 

I get 218 rows of data. But this should not be happening. I think some of the employees may have made mistakes. But errors are to be expected in data of this scale.

We can advice the Pres to appoint an Auditor to check some of the data independently, and make sure we have the right information.

5. Investigate pollution issues:
Contamination/pollution data was recorded for well sources. Lets have a look at the table

In [11]:
%%sql

SELECT
    *
FROM 
    well_pollution
LIMIT 10;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
10 rows affected.


[('KiRu28935224', datetime.datetime(2021, 1, 4, 9, 17), 'Bacteria: Giardia Lamblia', 0.0, 495.898, 'Contaminated: Biological'),
 ('AkLu01628224', datetime.datetime(2021, 1, 4, 9, 53), 'Bacteria: E. coli', 0.0, 6.09608, 'Contaminated: Biological'),
 ('HaZa21742224', datetime.datetime(2021, 1, 4, 10, 37), 'Inorganic contaminants: Zinc, Zinc, Lead, Cadmium', 2.715, 0.0, 'Contaminated: Chemical'),
 ('HaRu19725224', datetime.datetime(2021, 1, 4, 11, 4), 'Clean', 0.0288593, 9.56996e-05, 'Clean'),
 ('SoRu35703224', datetime.datetime(2021, 1, 4, 11, 29), 'Bacteria: E. coli', 0.0, 22.5009, 'Contaminated: Biological'),
 ('AkHa00070224', datetime.datetime(2021, 1, 4, 11, 42), 'Inorganic contaminants: Cadmium', 5.46739, 0.0, 'Contaminated: Chemical'),
 ('HaSe21346224', datetime.datetime(2021, 1, 4, 11, 52), 'Clean', 0.0140376, 8.98989e-05, 'Clean'),
 ('HaYa21468224', datetime.datetime(2021, 1, 4, 12, 3), 'Inorganic contaminants: Chromium, Barium, Chromium, Lead', 6.05137, 0.0, 'Contaminated: Chemical'),
 ('SoRu36278224', datetime.datetime(2021, 1, 4, 12, 24), 'Parasite: Cryptosporidium', 0.0, 485.162, 'Contaminated: Biological'),
 ('AkLu02155224', datetime.datetime(2021, 1, 4, 12, 29), 'Inorganic contaminants: Selenium, Arsenic', 7.64106, 0.0, 'Contaminated: Chemical')]

It looks like our scientists diligently recorded the water quality of all the wells. Some are contaminated with biological contaminants, while others are polluted with an excess of heavy metals and other pollutants. Based on the results, each well was classified as Clean, Contaminated: Biological or Contaminated: Chemical. It is important to know this because wells that are polluted with bio or other contaminants are not safe to drink. It looks like they recorded the source_id of each test, so we can link it to a source, at some place in Maji Ndogo.

In the well pollution table, the descriptions are notes taken by our scientists as text, so it will be challenging to process it. The
biological column is in units of CFU/mL, so it measures how much contamination is in the water. 0 is clean, and anything more than
0.01 is contaminated.
Let's check the integrity of the data. The worst case is if we have contamination, but we think we don't. People can get sick, so we
need to make sure there are no errors here.

In [12]:
%%sql

# Writing a query that checks if the results is Clean but the biological column is > 0.01.

SELECT 
    *
FROM 
    well_pollution
WHERE 
    results = "Clean"
    AND biological > 0.01;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
64 rows affected.


[('AkRu08936224', datetime.datetime(2021, 1, 8, 9, 22), 'Bacteria: E. coli', 0.0406458, 35.0068, 'Clean'),
 ('AkRu06489224', datetime.datetime(2021, 1, 10, 9, 44), 'Clean Bacteria: Giardia Lamblia', 0.0897904, 38.467, 'Clean'),
 ('SoRu38011224', datetime.datetime(2021, 1, 14, 15, 35), 'Bacteria: E. coli', 0.0425095, 19.2897, 'Clean'),
 ('AkKi00955224', datetime.datetime(2021, 1, 22, 12, 47), 'Bacteria: E. coli', 0.0812092, 40.2273, 'Clean'),
 ('KiHa22929224', datetime.datetime(2021, 2, 6, 13, 54), 'Bacteria: E. coli', 0.0722537, 18.4482, 'Clean'),
 ('KiRu25473224', datetime.datetime(2021, 2, 7, 15, 51), 'Clean Bacteria: Giardia Lamblia', 0.0630094, 24.4536, 'Clean'),
 ('HaRu17401224', datetime.datetime(2021, 3, 1, 13, 44), 'Clean Bacteria: Giardia Lamblia', 0.0649209, 25.8129, 'Clean'),
 ('AkRu07137224', datetime.datetime(2021, 3, 4, 13, 41), 'Clean Bacteria: Giardia Lamblia', 0.0656843, 18.2978, 'Clean'),
 ('KiRu27205224', datetime.datetime(2021, 3, 13, 14, 17), 'Clean Bacteria: Giardia Lamblia', 0.0418018, 49.4281, 'Clean'),
 ('AkLu02307224', datetime.datetime(2021, 3, 13, 15, 41), 'Bacteria: E. coli', 0.0709682, 35.203, 'Clean'),
 ('AkHa00514224', datetime.datetime(2021, 4, 11, 12, 11), 'Clean Bacteria: Giardia Lamblia', 0.0305404, 22.0255, 'Clean'),
 ('AmAb09197224', datetime.datetime(2021, 5, 10, 14, 23), 'Bacteria: E. coli', 0.0738421, 29.5398, 'Clean'),
 ('AmAm09776224', datetime.datetime(2021, 5, 23, 11, 28), 'Clean Bacteria: Giardia Lamblia', 0.0963821, 13.6574, 'Clean'),
 ('AkRu08599224', datetime.datetime(2021, 6, 24, 9, 36), 'Bacteria: E. coli', 0.0202987, 17.1466, 'Clean'),
 ('SoIl32894224', datetime.datetime(2021, 7, 11, 11, 37), 'Clean Bacteria: Giardia Lamblia', 0.0712408, 5.44957, 'Clean'),
 ('AkRu07366224', datetime.datetime(2021, 7, 23, 11, 19), 'Clean Bacteria: Giardia Lamblia', 0.0969458, 26.0308, 'Clean'),
 ('AkKi01423224', datetime.datetime(2021, 7, 30, 14, 57), 'Bacteria: E. coli', 0.012334, 4.75647, 'Clean'),
 ('AkRu08367224', datetime.datetime(2021, 8, 7, 14, 52), 'Bacteria: E. coli', 0.0498854, 41.4135, 'Clean'),
 ('KiHa23443224', datetime.datetime(2021, 9, 5, 12, 34), 'Clean Bacteria: Giardia Lamblia', 0.0828, 13.7162, 'Clean'),
 ('HaSe21137224', datetime.datetime(2021, 9, 20, 16, 57), 'Bacteria: E. coli', 0.0988156, 49.2078, 'Clean'),
 ('KiRu29323224', datetime.datetime(2021, 9, 25, 10, 1), 'Bacteria: E. coli', 0.0605767, 18.7618, 'Clean'),
 ('AkKi00919224', datetime.datetime(2021, 10, 30, 9, 3), 'Bacteria: E. coli', 0.0510793, 4.52353, 'Clean'),
 ('AkRu06340224', datetime.datetime(2021, 11, 1, 15, 24), 'Clean Bacteria: Giardia Lamblia', 0.051746, 36.4594, 'Clean'),
 ('SoBa31600224', datetime.datetime(2021, 11, 7, 14, 44), 'Bacteria: E. coli', 0.0687197, 4.49928, 'Clean'),
 ('AkRu06363224', datetime.datetime(2021, 11, 15, 13, 36), 'Clean Bacteria: Giardia Lamblia', 0.0882158, 45.3731, 'Clean'),
 ('AkRu08749224', datetime.datetime(2021, 11, 19, 13, 37), 'Clean Bacteria: E. coli', 0.0888194, 9.28085, 'Clean'),
 ('AkRu07066224', datetime.datetime(2021, 12, 10, 14, 40), 'Clean Bacteria: E. coli', 0.0151035, 44.4652, 'Clean'),
 ('KiAm21996224', datetime.datetime(2021, 12, 16, 10, 5), 'Clean Bacteria: E. coli', 0.0480415, 25.4174, 'Clean'),
 ('SoRu36080224', datetime.datetime(2021, 12, 31, 13, 57), 'Bacteria: E. coli', 0.0924292, 11.9485, 'Clean'),
 ('AkRu05377224', datetime.datetime(2022, 1, 1, 10, 53), 'Clean Bacteria: E. coli', 0.0226614, 0.613503, 'Clean'),
 ('HaRu20765224', datetime.datetime(2022, 2, 7, 12, 39), 'Clean Bacteria: E. coli', 0.0213831, 6.95778, 'Clean'),
 ('AmRu13112224', datetime.datetime(2022, 2, 13, 14, 57), 'Bacteria: E. coli', 0.0832901, 24.1679, 'Clean'),
 ('SoMa34473224', datetime.datetime(2022, 3, 10, 9, 1), 'Clean Bacteria: E. coli', 0.0745784, 13.6458, 'Clean'),
 ('SoKo33474224', datetime.datetime(2022, 3, 20, 16, 7), 'Clean Bacteria: E. coli', 0.0687177, 10.5712, 'Clean'),
 ('AkLu01944224', datetime.datetime(2022, 3, 24, 12, 23), 'Bacteria: E. coli', 0.0289, 6.75924, 'Clean'

If we compare the results of this query to the entire table it seems like we have some inconsistencies in how the well statuses are
recorded. Specifically, it seems that some data input personnel might have mistaken the description field for determining the clean-
liness of the water.

It seems like, in some cases, if the description field begins with the word “Clean”, the results have been classified as “Clean” in the results column, even though the biological column is > 0.01.

When we work with real-world data we may find inconsistencies due to data being misinterpreted based on a description rather than its
actual values. Let’s dive deeper into the cause of the issue with the biological contamination data.

One of the scientist mentioned that the descriptions should only have the word “Clean” if there is no biological contamination (and no chemical pollutants). Some data personnel must have copied the data from the scientist's notes into our database incorrectly. We need to find and remove the “Clean” part from all the descriptions that do have a biological contamination so this mistake is not made again.

The second issue has arisen from this error, but it is much more problematic. Some of the field surveyors have marked wells as Clean in
the results column because the description had the word “Clean” in it, even though they have a biological contamination. So we need
to find all the results that have a value greater than 0.01 in the biological column and have been set to Clean in the results column.

First, let's look at the descriptions. We need to identify the records that mistakenly have the word Clean in the description. However, it is important to remember that not all of our field surveyors used the description to set the results – some checked the actual data.

In [13]:
%%sql

SELECT
    *
FROM 
    well_pollution
WHERE 
    description LIKE "Clean%_"

 * mysql+pymysql://root:***@localhost:3306/md_water_services
38 rows affected.


[('AkRu06489224', datetime.datetime(2021, 1, 10, 9, 44), 'Clean Bacteria: Giardia Lamblia', 0.0897904, 38.467, 'Clean'),
 ('KiRu25473224', datetime.datetime(2021, 2, 7, 15, 51), 'Clean Bacteria: Giardia Lamblia', 0.0630094, 24.4536, 'Clean'),
 ('HaRu17401224', datetime.datetime(2021, 3, 1, 13, 44), 'Clean Bacteria: Giardia Lamblia', 0.0649209, 25.8129, 'Clean'),
 ('AkRu07137224', datetime.datetime(2021, 3, 4, 13, 41), 'Clean Bacteria: Giardia Lamblia', 0.0656843, 18.2978, 'Clean'),
 ('KiRu27205224', datetime.datetime(2021, 3, 13, 14, 17), 'Clean Bacteria: Giardia Lamblia', 0.0418018, 49.4281, 'Clean'),
 ('AkHa00514224', datetime.datetime(2021, 4, 11, 12, 11), 'Clean Bacteria: Giardia Lamblia', 0.0305404, 22.0255, 'Clean'),
 ('AmAm09776224', datetime.datetime(2021, 5, 23, 11, 28), 'Clean Bacteria: Giardia Lamblia', 0.0963821, 13.6574, 'Clean'),
 ('SoIl32894224', datetime.datetime(2021, 7, 11, 11, 37), 'Clean Bacteria: Giardia Lamblia', 0.0712408, 5.44957, 'Clean'),
 ('AkRu07366224', datetime.datetime(2021, 7, 23, 11, 19), 'Clean Bacteria: Giardia Lamblia', 0.0969458, 26.0308, 'Clean'),
 ('KiHa23443224', datetime.datetime(2021, 9, 5, 12, 34), 'Clean Bacteria: Giardia Lamblia', 0.0828, 13.7162, 'Clean'),
 ('AkRu06340224', datetime.datetime(2021, 11, 1, 15, 24), 'Clean Bacteria: Giardia Lamblia', 0.051746, 36.4594, 'Clean'),
 ('AkRu06363224', datetime.datetime(2021, 11, 15, 13, 36), 'Clean Bacteria: Giardia Lamblia', 0.0882158, 45.3731, 'Clean'),
 ('AkRu08749224', datetime.datetime(2021, 11, 19, 13, 37), 'Clean Bacteria: E. coli', 0.0888194, 9.28085, 'Clean'),
 ('AkRu07066224', datetime.datetime(2021, 12, 10, 14, 40), 'Clean Bacteria: E. coli', 0.0151035, 44.4652, 'Clean'),
 ('KiAm21996224', datetime.datetime(2021, 12, 16, 10, 5), 'Clean Bacteria: E. coli', 0.0480415, 25.4174, 'Clean'),
 ('AkRu05377224', datetime.datetime(2022, 1, 1, 10, 53), 'Clean Bacteria: E. coli', 0.0226614, 0.613503, 'Clean'),
 ('HaRu20765224', datetime.datetime(2022, 2, 7, 12, 39), 'Clean Bacteria: E. coli', 0.0213831, 6.95778, 'Clean'),
 ('SoMa34473224', datetime.datetime(2022, 3, 10, 9, 1), 'Clean Bacteria: E. coli', 0.0745784, 13.6458, 'Clean'),
 ('SoKo33474224', datetime.datetime(2022, 3, 20, 16, 7), 'Clean Bacteria: E. coli', 0.0687177, 10.5712, 'Clean'),
 ('AkHa00226224', datetime.datetime(2022, 3, 26, 11, 53), 'Clean Bacteria: E. coli', 0.0424571, 39.4978, 'Clean'),
 ('AkRu06881224', datetime.datetime(2022, 3, 28, 13, 37), 'Clean Bacteria: E. coli', 0.0202831, 33.0692, 'Clean'),
 ('AkLu02254224', datetime.datetime(2022, 4, 25, 12, 51), 'Clean Bacteria: E. coli', 0.0635533, 16.9482, 'Clean'),
 ('AkRu08416224', datetime.datetime(2022, 5, 29, 15, 59), 'Clean Bacteria: E. coli', 0.0406701, 27.0805, 'Clean'),
 ('SoMa34205224', datetime.datetime(2022, 7, 31, 13, 25), 'Clean Bacteria: E. coli', 0.0726217, 23.892, 'Clean'),
 ('KiIs23482224', datetime.datetime(2022, 8, 15, 16, 41), 'Clean Bacteria: E. coli', 0.0539137, 41.0842, 'Clean'),
 ('AmRu13252224', datetime.datetime(2022, 8, 20, 14, 23), 'Clean Bacteria: E. coli', 0.0653589, 30.8643, 'Clean'),
 ('KiRu27297224', datetime.datetime(2022, 11, 21, 13, 18), 'Clean Bacteria: E. coli', 0.0856842, 8.62509, 'Clean'),
 ('KiRu26929224', datetime.datetime(2022, 12, 18, 10, 52), 'Clean Bacteria: E. coli', 0.0748808, 26.6541, 'Clean'),
 ('HaAm16042224', datetime.datetime(2022, 12, 22, 13, 52), 'Clean Bacteria: E. coli', 0.0605767, 4.54755, 'Clean'),
 ('AkRu03107224', datetime.datetime(2023, 1, 1, 15, 37), 'Clean Bacteria: E. coli', 0.0901101, 18.3118, 'Clean'),
 ('KiHa23111224', datetime.datetime(2023, 1, 12, 15, 56), 'Clean Bacteria: E. coli', 0.0161869, 36.8748, 'Clean'),
 ('AkRu03771224', datetime.datetime(2023, 2, 26, 12, 40), 'Clean Bacteria: E. coli', 0.0625017, 19.7648, 'Clean'),
 ('AkRu04253224', datetime.datetime(2023, 2, 26, 13, 46), 'Clean Bacteria: E. coli', 0.0273811, 3.08854, 'Clean'),
 ('HaSe20949224', datetime.datetime(2023, 4, 3, 16, 26), 'Clean Bacteria: E. coli', 0.0458311, 32.9004, 'Clean'),
 ('Ki

The query returns 38 wrong descriptions.


Now we need to fix these descriptions so that we don’t encounter this issue again in the future.


Looking at the results we can see two different descriptions that we need to fix:
1. All records that mistakenly have Clean Bacteria: E. coli should updated to Bacteria: E. coli
2. All records that mistakenly have Clean Bacteria: Giardia Lamblia should updated to Bacteria: Giardia Lamblia


The second issue we need to fix is in our results column. We need to update the results column from Clean to Contaminated: Biological where the biological column has a value greater than 0.01.

In [14]:
%%sql

# Now, when we change any data on the database, we need to be SURE there are no errors, as this could fill the database with 
# incorrect values. A safer way to do the UPDATE is by testing the changes on a copy of the table first.

CREATE TABLE
    md_water_services.well_pollution_copy
AS (
    SELECT
        *
    FROM
        md_water_services.well_pollution
    );

 * mysql+pymysql://root:***@localhost:3306/md_water_services
(pymysql.err.OperationalError) (1050, "Table 'well_pollution_copy' already exists")
[SQL: # Now, when we change any data on the database, we need to be SURE there are no errors, as this could fill the database with 
# incorrect values. A safer way to do the UPDATE is by testing the changes on a copy of the table first.

CREATE TABLE
    md_water_services.well_pollution_copy
AS (
    SELECT
        *
    FROM
        md_water_services.well_pollution
    );]
(Background on this error at: http://sqlalche.me/e/14/e3q8)


In [ ]:
%%sql

# Case 1a: Update descriptions that mistakenly mention `Clean Bacteria: E. coli` to `Bacteria: E. coli`
UPDATE 
    well_pollution_copy
SET
    description = 'Bacteria: E. coli'
WHERE
    description = 'Clean Bacteria: E. coli';
# Case 1b: Update the descriptions that mistakenly mention `Clean Bacteria: Giardia Lamblia` to `Bacteria: Giardia Lamblia
UPDATE 
    well_pollution_copy
SET 
    description = 'Bacteria: Giardia Lamblia'
WHERE 
    description = 'Clean Bacteria: Giardia Lamblia';
# Case 2: Update the `result` to `Contaminated: Biological` where `biological` is greater than 0.01 plus current results is `Clean`
UPDATE 
    well_pollution_copy
SET 
    result = 'Contaminated: Biological'
WHERE 
    biological > 0.01 AND results = 'Clean'

In [15]:
%%sql

# checking if our errors are fixed

SELECT
    *
FROM
    well_pollution_copy
WHERE
    description LIKE "Clean_%"
    OR (results = "Clean" AND biological > 0.01);

 * mysql+pymysql://root:***@localhost:3306/md_water_services
0 rows affected.


[]

0 results. We can now see that the the update was successful.
We can choose to now update it on the original well pollution table, but I choose to leave it on the copy table and do future queries from it

Now Lets answer the MCQs below

In [16]:
%%sql

# Exploring the employee table

SELECT 
    *
FROM 
    employee
LIMIT 5;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[(0, 'Amara Jengo', '+99637993287', 'amara.jengo@ndogowater.gov', '36 Pwani Mchangani Road', 'Sokoto', 'Ilanga', 'Field Surveyor'),
 (1, 'Bello Azibo', '+99643864786', 'bello.azibo@ndogowater.gov', '129 Ziwa La Kioo Road', 'Kilimani', 'Rural', 'Field Surveyor'),
 (2, 'Bakari Iniko', '+99222599041', 'bakari.iniko@ndogowater.gov', '18 Mlima Tazama Avenue', 'Hawassa', 'Rural', 'Field Surveyor'),
 (3, 'Malachi Mavuso', '+99945849900', 'malachi.mavuso@ndogowater.gov', '100 Mogadishu Road', 'Akatsi', 'Lusaka', 'Field Surveyor'),
 (4, 'Cheche Buhle', '+99381679640', 'cheche.buhle@ndogowater.gov', '1 Savanna Street', 'Akatsi', 'Rural', 'Field Surveyor')]

Anwering some question to fetch out some data

In [17]:
%%sql

# Q1. What is the address of Bello Azibo?

SELECT 
    address
FROM 
    employee
WHERE 
    employee_name = 'Bello Azibo';

 * mysql+pymysql://root:***@localhost:3306/md_water_services
1 rows affected.


[('129 Ziwa La Kioo Road',)]

In [18]:
%%sql

# Q2. What is the name and phone number of our microbiologist?

SELECT 
    employee_name,
    phone_number
FROM 
    employee
WHERE 
    position = 'Micro Biologist'

 * mysql+pymysql://root:***@localhost:3306/md_water_services
1 rows affected.


[('Vuyisile Ghadir', '+99712584936')]

In [19]:
%%sql

# -- Q.3 What is the source_id of the water source shared by the most number of people? Hint: Use a comparison operator.
# From the result, the source_id is AkRu05603224

SELECT 
    *
FROM 
    water_source 
ORDER BY 
    number_of_people_served DESC 
LIMIT 5;



 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[('AkRu05603224', 'shared_tap', 3998),
 ('HaRu19509224', 'shared_tap', 3998),
 ('AmAs10911224', 'shared_tap', 3996),
 ('AkRu04862224', 'shared_tap', 3996),
 ('KiHa22867224', 'shared_tap', 3996)]

In [20]:
%%sql 

# Q4. What is the population of Maji Ndogo? Hint: Start by searching the data_dictionary table for the word 'population'.

SELECT 
    * 
FROM 
    data_dictionary 
WHERE description LIKE '%population%';

 * mysql+pymysql://root:***@localhost:3306/md_water_services
2 rows affected.


[('global_water_access', 'pop_n', 'The national population size estimate in thousands', 'float', ''),
 ('global_water_access', 'pop_u', 'The urban population share estimate in percentage points (%)', 'float', '')]

In [21]:
%%sql 

# now I can tell where the population is in the database. Its in the global_water_access table and the column name is 'pop_n'.
#  And its estimated in 1000s ie; n*1000, where n is the estimated figure. From this info I can fetch the population of maji_ndogo
# Hence population of Maji Ndogo is 27,628,100

SELECT 
    *
FROM 
    global_water_access
WHERE 
    name = 'Maji Ndogo';



 * mysql+pymysql://root:***@localhost:3306/md_water_services
2 rows affected.


[('Maji Ndogo', 'Sub-Saharan Africa', 2015, 27628.1, 36.0, 47.6, 31.4, 12.8, 8.2, 41.15, 36.68, 12.67, 9.49, 58.32, 22.57, 12.21, 6.89),
 ('Maji Ndogo', 'Sub-Saharan Africa', 2020, 27628.1, 36.0, 47.6, 31.4, 12.8, 8.2, 41.15, 36.68, 12.67, 9.49, 58.32, 22.57, 12.21, 6.89)]

In [22]:
%%sql

# Q5. Which SQL query returns records of employees who are Civil Engineers residing in Dahabu or living on an avenue?

SELECT 
    *
FROM 
    employee
WHERE 
    position = 'Civil Engineer'
    AND (province_name = 'Dahabu' OR address LIKE '%Avenue%');

 * mysql+pymysql://root:***@localhost:3306/md_water_services
2 rows affected.


[(71, 'Jengo Rudo', '+99317854629', 'jengo.rudo@ndogowater.gov', '33 Angélique Kidjo Avenue', 'Amanzi', 'Dahabu', 'Civil Engineer'),
 (81, 'Nia Vuyisile', '+99273841596', 'nia.vuyisile@ndogowater.gov', '52 Moroni Avenue', 'Sokoto', 'Rural', 'Civil Engineer')]

In [23]:
%%sql 

#Q6. Create a query to identify potentially suspicious field workers based on an anonymous tip. This is the description we are given:

#The employee’s phone number contained the digits 86 or 11. 
#The employee’s last name started with either an A or an M. 
#The employee was a Field Surveyor.

SELECT 
    employee_name
FROM 
    employee 
WHERE 
     (phone_number LIKE '%86%'
    OR phone_number LIKE '%11%')
    AND (employee_name LIKE '% A%' 
    OR employee_name LIKE '% M%')
    AND position = 'Field Surveyor';

 * mysql+pymysql://root:***@localhost:3306/md_water_services
2 rows affected.


[('Bello Azibo',), ('Zuriel Matembo',)]

In [24]:
%%sql 

#Q7. What is the result of the following query? Choose the most appropriate description of the results set.

SELECT 
    *
FROM 
    well_pollution
WHERE 
    (description LIKE 'Clean_%' OR results = 'Clean') 
    AND biological < 0.01;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
4916 rows affected.


[('HaRu19725224', datetime.datetime(2021, 1, 4, 11, 4), 'Clean', 0.0288593, 9.56996e-05, 'Clean'),
 ('HaSe21346224', datetime.datetime(2021, 1, 4, 11, 52), 'Clean', 0.0140376, 8.98989e-05, 'Clean'),
 ('AkRu05973224', datetime.datetime(2021, 1, 4, 13, 44), 'Clean', 0.084344, 6.2396e-05, 'Clean'),
 ('AkHa00706224', datetime.datetime(2021, 1, 4, 16, 10), 'Clean', 0.0862577, 4.22834e-05, 'Clean'),
 ('HaRu20773224', datetime.datetime(2021, 1, 7, 9, 8), 'Clean', 0.0599682, 2.42289e-05, 'Clean'),
 ('AkRu04914224', datetime.datetime(2021, 1, 7, 11, 35), 'Clean', 0.050281, 9.42943e-05, 'Clean'),
 ('HaSe20931224', datetime.datetime(2021, 1, 7, 11, 46), 'Clean', 0.0224843, 9.87848e-05, 'Clean'),
 ('AmAm09579224', datetime.datetime(2021, 1, 7, 13, 2), 'Clean', 0.0830754, 1.10413e-05, 'Clean'),
 ('KiHa23405224', datetime.datetime(2021, 1, 7, 15, 29), 'Clean', 0.0633972, 5.88518e-05, 'Clean'),
 ('AmDa11574224', datetime.datetime(2021, 1, 7, 15, 51), 'Clean', 0.0290295, 6.11353e-05, 'Clean'),
 ('KiRu26713224', datetime.datetime(2021, 1, 7, 16, 52), 'Clean', 0.0854418, 2.91209e-05, 'Clean'),
 ('SoMa34243224', datetime.datetime(2021, 1, 8, 9, 17), 'Clean', 0.0708139, 6.21986e-05, 'Clean'),
 ('AmBe11186224', datetime.datetime(2021, 1, 8, 10, 24), 'Clean', 0.0761867, 3.15558e-05, 'Clean'),
 ('KiRu26783224', datetime.datetime(2021, 1, 8, 11, 52), 'Clean', 0.0340469, 8.68879e-05, 'Clean'),
 ('AkLu02295224', datetime.datetime(2021, 1, 8, 13, 31), 'Clean', 0.0209784, 3.97724e-05, 'Clean'),
 ('HaSe21277224', datetime.datetime(2021, 1, 9, 10, 46), 'Clean', 0.0391768, 3.81981e-05, 'Clean'),
 ('AkHa00744224', datetime.datetime(2021, 1, 9, 11, 14), 'Clean', 0.0370928, 7.16733e-05, 'Clean'),
 ('AkRu08868224', datetime.datetime(2021, 1, 9, 11, 51), 'Clean', 0.088961, 4.37723e-05, 'Clean'),
 ('HaSe20883224', datetime.datetime(2021, 1, 9, 12, 33), 'Clean', 0.0732487, 3.84149e-06, 'Clean'),
 ('AkRu06795224', datetime.datetime(2021, 1, 9, 12, 58), 'Clean', 0.0834844, 8.78906e-05, 'Clean'),
 ('AkRu07238224', datetime.datetime(2021, 1, 9, 13, 55), 'Clean', 0.0167354, 2.79285e-05, 'Clean'),
 ('KiRu30187224', datetime.datetime(2021, 1, 9, 14, 21), 'Clean', 0.0116701, 7.59707e-05, 'Clean'),
 ('AkRu04348224', datetime.datetime(2021, 1, 9, 14, 30), 'Clean', 0.0847516, 9.6068e-05, 'Clean'),
 ('KiRu26106224', datetime.datetime(2021, 1, 9, 14, 59), 'Clean', 0.0743687, 5.24281e-05, 'Clean'),
 ('KiIs24152224', datetime.datetime(2021, 1, 9, 16, 40), 'Clean', 0.0570679, 7.39365e-05, 'Clean'),
 ('AkRu06588224', datetime.datetime(2021, 1, 10, 10, 3), 'Clean', 0.082105, 4.0181e-05, 'Clean'),
 ('SoRu36464224', datetime.datetime(2021, 1, 10, 10, 15), 'Clean', 0.0231531, 6.37108e-05, 'Clean'),
 ('KiRu27574224', datetime.datetime(2021, 1, 10, 11, 41), 'Clean', 0.0356204, 9.80107e-05, 'Clean'),
 ('AkRu04804224', datetime.datetime(2021, 1, 10, 11, 50), 'Clean', 0.0716538, 9.89214e-05, 'Clean'),
 ('KiRu29881224', datetime.datetime(2021, 1, 10, 12, 8), 'Clean', 0.0739334, 5.74722e-07, 'Clean'),
 ('SoRu37976224', datetime.datetime(2021, 1, 10, 13, 29), 'Clean', 0.0766173, 6.10948e-06, 'Clean'),
 ('KiRu29286224', datetime.datetime(2021, 1, 10, 15, 11), 'Clean', 0.0645078, 2.88232e-05, 'Clean'),
 ('KiRu29107224', datetime.datetime(2021, 1, 10, 16, 10), 'Clean', 0.0806449, 2.57877e-05, 'Clean'),
 ('AkRu08423224', datetime.datetime(2021, 1, 10, 16, 49), 'Clean', 0.0335904, 4.24688e-05, 'Clean'),
 ('AmRu14471224', datetime.datetime(2021, 1, 11, 10, 7), 'Clean', 0.072325, 3.49809e-05, 'Clean'),
 ('HaSe21227224', datetime.datetime(2021, 1, 11, 10, 33), 'Clean', 0.0417335, 4.74981e-05, 'Clean'),
 ('AkRu07336224', datetime.datetime(2021, 1, 11, 10, 50), 'Clean', 0.0852829, 3.25477e-05, 'Clean'),
 ('AmRu15108224', datetime.datetime(2021, 1, 11, 11, 43), 'Clean', 0.0737846, 2.02442e-05, 'Clean'),
 ('AkLu01724224', datetime.datetime(2021, 1, 11, 11, 58), 'Clean', 0.0905594, 3.57788e-06, 'Clean'),
 ('AkRu08461224', datetime.datetime(2021, 1, 11, 12, 52), 'Clean', 0.0590685, 5.71569e-05, 'Clean'),
 

4916 records are returned. This query describes the pollution samples that had an insignificant amount of biological contamination.

In [25]:
%%sql 

# Q8. Which query will identify the records with a quality score of 10, visited more than once?

SELECT 
    *
FROM 
    water_quality 
WHERE 
    visit_count > 1
    AND subjective_quality_score = 10

 * mysql+pymysql://root:***@localhost:3306/md_water_services
1526 rows affected.


[(59, 10, 2),
 (67, 10, 3),
 (85, 10, 4),
 (128, 10, 5),
 (137, 10, 2),
 (232, 10, 3),
 (263, 10, 6),
 (269, 10, 2),
 (271, 10, 3),
 (317, 10, 4),
 (318, 10, 7),
 (358, 10, 4),
 (363, 10, 2),
 (378, 10, 2),
 (398, 10, 8),
 (431, 10, 5),
 (436, 10, 6),
 (439, 10, 5),
 (441, 10, 3),
 (450, 10, 3),
 (459, 10, 7),
 (460, 10, 4),
 (463, 10, 6),
 (470, 10, 4),
 (476, 10, 8),
 (488, 10, 5),
 (498, 10, 5),
 (508, 10, 7),
 (566, 10, 6),
 (618, 10, 2),
 (630, 10, 8),
 (632, 10, 6),
 (656, 10, 3),
 (662, 10, 7),
 (730, 10, 7),
 (741, 10, 8),
 (749, 10, 8),
 (752, 10, 2),
 (758, 10, 4),
 (801, 10, 2),
 (817, 10, 5),
 (818, 10, 3),
 (819, 10, 2),
 (850, 10, 2),
 (869, 10, 3),
 (894, 10, 3),
 (904, 10, 4),
 (905, 10, 3),
 (911, 10, 6),
 (918, 10, 4),
 (921, 10, 7),
 (930, 10, 4),
 (932, 10, 5),
 (937, 10, 5),
 (941, 10, 4),
 (952, 10, 6),
 (955, 10, 6),
 (959, 10, 8),
 (989, 10, 5),
 (1017, 10, 5),
 (1020, 10, 6),
 (1022, 10, 7),
 (1055, 10, 6),
 (1076, 10, 7),
 (1102, 10, 7),
 (1109, 10, 7),
 (1143, 10, 8),
 (1153, 10, 8),
 (1176, 10, 2),
 (1184, 10, 3),
 (1185, 10, 8),
 (1263, 10, 8),
 (1333, 10, 4),
 (1372, 10, 2),
 (1374, 10, 5),
 (1382, 10, 6),
 (1392, 10, 3),
 (1404, 10, 4),
 (1430, 10, 7),
 (1492, 10, 5),
 (1496, 10, 2),
 (1499, 10, 8),
 (1612, 10, 3),
 (1656, 10, 6),
 (1669, 10, 4),
 (1739, 10, 7),
 (1746, 10, 2),
 (1753, 10, 2),
 (1755, 10, 3),
 (1760, 10, 5),
 (1805, 10, 8),
 (1840, 10, 4),
 (1841, 10, 3),
 (1843, 10, 6),
 (1860, 10, 4),
 (1875, 10, 5),
 (1877, 10, 7),
 (1883, 10, 5),
 (1885, 10, 8),
 (1886, 10, 6),
 (1988, 10, 7),
 (2056, 10, 6),
 (2067, 10, 8),
 (2124, 10, 7),
 (2135, 10, 8),
 (2298, 10, 2),
 (2316, 10, 3),
 (2334, 10, 4),
 (2342, 10, 5),
 (2445, 10, 6),
 (2559, 10, 7),
 (2657, 10, 8),
 (2702, 10, 2),
 (2780, 10, 3),
 (2783, 10, 4),
 (2820, 10, 5),
 (2901, 10, 6),
 (3047, 10, 7),
 (3091, 10, 8),
 (3203, 10, 2),
 (3229, 10, 3),
 (3288, 10, 4),
 (3404, 10, 5),
 (3477, 10, 6),
 (3489, 10, 2),
 (3550, 10, 7),
 (3562, 10, 3),
 (3590, 10, 4),
 (3636, 10, 8),
 (3702, 10, 5),
 (3718, 10, 6),
 (3793, 10, 7),
 (3835, 10, 8),
 (4750, 10, 2),
 (4827, 10, 3),
 (4873, 10, 4),
 (4906, 10, 2),
 (4970, 10, 3),
 (5048, 10, 5),
 (5049, 10, 6),
 (5052, 10, 4),
 (5068, 10, 2),
 (5072, 10, 2),
 (5084, 10, 3),
 (5087, 10, 5),
 (5089, 10, 7),
 (5096, 10, 3),
 (5098, 10, 8),
 (5118, 10, 6),
 (5146, 10, 4),
 (5165, 10, 4),
 (5180, 10, 7),
 (5181, 10, 5),
 (5221, 10, 5),
 (5262, 10, 6),
 (5266, 10, 6),
 (5279, 10, 2),
 (5308, 10, 8),
 (5310, 10, 7),
 (5343, 10, 2),
 (5345, 10, 7),
 (5381, 10, 3),
 (5391, 10, 4),
 (5392, 10, 8),
 (5410, 10, 8),
 (5441, 10, 3),
 (5469, 10, 5),
 (5470, 10, 4),
 (5493, 10, 5),
 (5502, 10, 6),
 (5508, 10, 7),
 (5567, 10, 6),
 (5604, 10, 7),
 (5621, 10, 2),
 (5639, 10, 8),
 (5656, 10, 2),
 (5673, 10, 3),
 (5743, 10, 8),
 (5754, 10, 4),
 (5774, 10, 3),
 (5822, 10, 5),
 (5846, 10, 4),
 (5907, 10, 6),
 (5916, 10, 5),
 (5918, 10, 2),
 (5925, 10, 6),
 (5931, 10, 3),
 (5937, 10, 7),
 (5946, 10, 4),
 (6009, 10, 8),
 (6023, 10, 7),
 (6048, 10, 5),
 (6066, 10, 8),
 (6074, 10, 2),
 (6160, 10, 3),
 (6195, 10, 6),
 (6226, 10, 4),
 (6288, 10, 7),
 (6335, 10, 8),
 (6342, 10, 5),
 (6397, 10, 6),
 (6409, 10, 7),
 (6456, 10, 8),
 (6620, 10, 2),
 (6690, 10, 3),
 (6731, 10, 2),
 (6732, 10, 3),
 (6826, 10, 4),
 (6830, 10, 5),
 (6841, 10, 4),
 (6853, 10, 6),
 (6864, 10, 5),
 (6909, 10, 7),
 (6930, 10, 6),
 (6989, 10, 8),
 (7018, 10, 7),
 (7097, 10, 8),
 (7251, 10, 2),
 (7266, 10, 3),
 (7271, 10, 2),
 (7284, 10, 2),
 (7294, 10, 3),
 (7296, 10, 4),
 (7303, 10, 3),
 (7345, 10, 5),
 (7373, 10, 4),
 (7397, 10, 4),
 (7400, 10, 5),
 (7424, 10, 5),
 (7466, 10, 6),
 (7495, 10, 6),
 (7524, 10, 7),
 (7552, 10, 6),
 (7582, 10, 8),
 (7594, 10, 7),
 (7631, 10, 7),
 (7679, 10, 8),
 (7711, 10, 2),
 (7715, 10, 8),
 (7747, 10, 2),
 (7752, 10, 3),
 (7761, 10, 2),
 (7778, 10, 3),
 (7779, 10, 4),
 (7781, 10, 4),
 (7794, 10, 3),
 (7809, 10, 5),
 (7830, 10, 4),
 (7880, 10, 5),
 (7894, 10, 5),
 (7917, 10, 6),
 (7921, 10, 6),
 (7980, 10, 2),
 (8029, 10, 7)

In [26]:
%%sql 

#Updating Bello Azibo's number with his correct number '+99643864786'

UPDATE employee
SET phone_number = '+99643864786'
WHERE employee_name = 'Bello Azibo';



 * mysql+pymysql://root:***@localhost:3306/md_water_services
1 rows affected.


ResourceClosedError: This result object does not return rows. It has been closed automatically.

In [27]:
%%sql 

SELECT employee_name, phone_number
FROM employee
WHERE employee_name = 'Bello Azibo';

 * mysql+pymysql://root:***@localhost:3306/md_water_services
1 rows affected.


[('Bello Azibo', '+99643864786')]

In [31]:
%%sql 

# Q10. How many rows of data are returned for the following query?


SELECT 
    * 
FROM 
    well_pollution_copy
WHERE 
    description IN ('Parasite: Cryptosporidium', 'biologically contaminated')
    OR (results = 'Clean' AND biological > 0.01)

 * mysql+pymysql://root:***@localhost:3306/md_water_services
570 rows affected.


[('SoRu36278224', datetime.datetime(2021, 1, 4, 12, 24), 'Parasite: Cryptosporidium', 0.0, 485.162, 'Contaminated: Biological'),
 ('HaDe16499224', datetime.datetime(2021, 1, 7, 12, 53), 'Parasite: Cryptosporidium', 0.0, 485.162, 'Contaminated: Biological'),
 ('KiRu26584224', datetime.datetime(2021, 1, 8, 13, 10), 'Parasite: Cryptosporidium', 0.0, 485.162, 'Contaminated: Biological'),
 ('KiZu31025224', datetime.datetime(2021, 1, 10, 14, 36), 'Parasite: Cryptosporidium', 0.0, 485.162, 'Contaminated: Biological'),
 ('HaRu17300224', datetime.datetime(2021, 1, 16, 16, 47), 'Parasite: Cryptosporidium', 0.0, 485.162, 'Contaminated: Biological'),
 ('SoRu35543224', datetime.datetime(2021, 1, 18, 9, 49), 'Parasite: Cryptosporidium', 0.0, 485.162, 'Contaminated: Biological'),
 ('KiRu26626224', datetime.datetime(2021, 1, 23, 10, 27), 'Parasite: Cryptosporidium', 0.0, 485.162, 'Contaminated: Biological'),
 ('HaRu19424224', datetime.datetime(2021, 1, 29, 14, 15), 'Parasite: Cryptosporidium', 0.0, 485.162, 'Contaminated: Biological'),
 ('KiAm22282224', datetime.datetime(2021, 1, 30, 16, 28), 'Parasite: Cryptosporidium', 0.0, 485.162, 'Contaminated: Biological'),
 ('KiRu26616224', datetime.datetime(2021, 2, 5, 13, 48), 'Parasite: Cryptosporidium', 0.0, 485.162, 'Contaminated: Biological'),
 ('HaRu20737224', datetime.datetime(2021, 2, 6, 11, 21), 'Parasite: Cryptosporidium', 0.0, 485.162, 'Contaminated: Biological'),
 ('AmRu14064224', datetime.datetime(2021, 2, 6, 11, 44), 'Parasite: Cryptosporidium', 0.0, 485.162, 'Contaminated: Biological'),
 ('SoRu38854224', datetime.datetime(2021, 2, 6, 15, 42), 'Parasite: Cryptosporidium', 0.0, 485.162, 'Contaminated: Biological'),
 ('SoRu37692224', datetime.datetime(2021, 2, 7, 15, 29), 'Parasite: Cryptosporidium', 0.0, 485.162, 'Contaminated: Biological'),
 ('SoMa34595224', datetime.datetime(2021, 2, 12, 12, 58), 'Parasite: Cryptosporidium', 0.0, 485.162, 'Contaminated: Biological'),
 ('KiMr25029224', datetime.datetime(2021, 2, 15, 14, 7), 'Parasite: Cryptosporidium', 0.0, 485.162, 'Contaminated: Biological'),
 ('SoRu35481224', datetime.datetime(2021, 2, 15, 16, 17), 'Parasite: Cryptosporidium', 0.0, 485.162, 'Contaminated: Biological'),
 ('AmRu14810224', datetime.datetime(2021, 2, 18, 10, 56), 'Parasite: Cryptosporidium', 0.0, 485.162, 'Contaminated: Biological'),
 ('SoIl32401224', datetime.datetime(2021, 2, 19, 15, 2), 'Parasite: Cryptosporidium', 0.0, 485.162, 'Contaminated: Biological'),
 ('AkRu02912224', datetime.datetime(2021, 2, 22, 13, 24), 'Parasite: Cryptosporidium', 0.0, 485.162, 'Contaminated: Biological'),
 ('KiRu25945224', datetime.datetime(2021, 2, 22, 14, 48), 'Parasite: Cryptosporidium', 0.0, 485.162, 'Contaminated: Biological'),
 ('AkRu07694224', datetime.datetime(2021, 2, 26, 12, 31), 'Parasite: Cryptosporidium', 0.0, 485.162, 'Contaminated: Biological'),
 ('HaDe16409224', datetime.datetime(2021, 2, 28, 10, 46), 'Parasite: Cryptosporidium', 0.0, 485.162, 'Contaminated: Biological'),
 ('HaRu19028224', datetime.datetime(2021, 3, 1, 13, 18), 'Parasite: Cryptosporidium', 0.0, 485.162, 'Contaminated: Biological'),
 ('KiRu26863224', datetime.datetime(2021, 3, 4, 12, 52), 'Parasite: Cryptosporidium', 0.0, 485.162, 'Contaminated: Biological'),
 ('HaRu17436224', datetime.datetime(2021, 3, 5, 10, 39), 'Parasite: Cryptosporidium', 0.0, 485.162, 'Contaminated: Biological'),
 ('KiRu29165224', datetime.datetime(2021, 3, 8, 12, 3), 'Parasite: Cryptosporidium', 0.0, 485.162, 'Contaminated: Biological'),
 ('AmRu15748224', datetime.datetime(2021, 3, 11, 15, 1), 'Parasite: Cryptosporidium', 0.0, 485.162, 'Contaminated: Biological'),
 ('SoRu38543224', datetime.datetime(2021, 3, 11, 15, 25), 'Parasite: Cryptosporidium', 0.0, 485.162, 'Contaminated: Biological'),
 ('SoBa31750224', datetime.datetime(2021, 3, 13, 9, 6), 'Parasite: Cryptosporidium', 0.0, 485.162, 'Contaminated: Biological'),
 ('KiRu28517224', datetime.datetime(2021, 3, 13, 16, 36), 'Parasite: Cryptosporidium', 0.0, 485.162, 'Contaminated: Bio